File: go/containment_test.go
package jostraca

import (
	"errors"
	"sort"
	"strings"
	"testing"
)

// Regression tests for output-path containment and backup-file naming,
// mirroring ts/test/containment.test.ts.

const containWhen int64 = 1735689600000

func outKeys(m *MemFS) []string {
	keys := []string{}
	for k := range m.Vol() {
		if !strings.Contains(k, ".jostraca") {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)
	return keys
}

func eqKeys(t *testing.T, got, want []string) {
	t.Helper()
	if len(got) != len(want) {
		t.Fatalf("keys: got %v want %v", got, want)
	}
	for i := range got {
		if got[i] != want[i] {
			t.Fatalf("keys: got %v want %v", got, want)
		}
	}
}

func containJ(m *MemFS, extra ...Option) *J {
	o := []Option{WithFS(m), WithFolder("/out"), WithNow(func() int64 { return containWhen })}
	return New(append(o, extra...)...)
}

// A File with no enclosing Project must stay under the output folder.
// (The Go port already resolved this correctly; the test locks it in so
// the TS fix and the port cannot drift apart again.)
func TestFileWithoutProjectStaysInOutputFolder(t *testing.T) {
	m := NewMemFS()
	if _, err := containJ(m).Generate(Options{}, func(j *J) {
		j.File("x.txt", func(j *J) { j.Content("hi\n") })
	}); err != nil {
		t.Fatal(err)
	}
	eqKeys(t, outKeys(m), []string{"/out/x.txt"})
}

func TestFolderWithoutProjectStaysInOutputFolder(t *testing.T) {
	m := NewMemFS()
	if _, err := containJ(m).Generate(Options{}, func(j *J) {
		j.Folder("sub", func(j *J) {
			j.File("y.txt", func(j *J) { j.Content("hi\n") })
		})
	}); err != nil {
		t.Fatal(err)
	}
	eqKeys(t, outKeys(m), []string{"/out/sub/y.txt"})
}

func TestFileNameRejectsTraversal(t *testing.T) {
	m := NewMemFS()
	_, err := containJ(m).Generate(Options{}, func(j *J) {
		j.Project(ProjectProps{Folder: "p"}, func(j *J) {
			j.File("../../../../etc/pwned.txt", func(j *J) { j.Content("OWNED\n") })
		})
	})
	if !errors.Is(err, ErrNameTraversal) {
		t.Fatalf("expected ErrNameTraversal, got %v", err)
	}
	if len(outKeys(m)) != 0 {
		t.Fatalf("no file should have been written, got %v", outKeys(m))
	}
}

func TestFolderNameRejectsTraversal(t *testing.T) {
	m := NewMemFS()
	_, err := containJ(m).Generate(Options{}, func(j *J) {
		j.Project(ProjectProps{Folder: "p"}, func(j *J) {
			j.Folder("../..", func(j *J) {
				j.File("e.txt", func(j *J) { j.Content("X\n") })
			})
		})
	})
	if !errors.Is(err, ErrNameTraversal) {
		t.Fatalf("expected ErrNameTraversal, got %v", err)
	}
}

// A leading "/" in a Folder name composes with the Project folder and
// must keep working — see the absolute_paths parity scenario.
func TestAbsoluteFolderNameStillComposes(t *testing.T) {
	m := NewMemFS()
	j := New(WithFS(m), WithFolder("/top"), WithNow(func() int64 { return containWhen }))
	if _, err := j.Generate(Options{}, func(j *J) {
		j.Project(ProjectProps{Folder: "/top/sdk"}, func(j *J) {
			j.Folder("/code/js", func(j *J) {
				j.File("foo.js", func(j *J) { j.Content("// foo\n") })
			})
		})
	}); err != nil {
		t.Fatal(err)
	}
	eqKeys(t, outKeys(m), []string{"/top/sdk/code/js/foo.js"})
}

// Every dotfile in a folder must get its own backup path.
func TestPreserveBacksUpDotfilesDistinctly(t *testing.T) {
	m := NewMemFS()
	_ = m.WriteFile("/out/p/.env", []byte("OLD-ENV\n"))
	_ = m.WriteFile("/out/p/.npmrc", []byte("OLD-NPMRC\n"))

	tr := true
	j := containJ(m, WithExisting(Existing{Txt: ExistingTxt{Preserve: &tr}}))
	if _, err := j.Generate(Options{}, func(j *J) {
		j.Project(ProjectProps{Folder: "p"}, func(j *J) {
			j.File(".env", func(j *J) { j.Content("NEW-ENV\n") })
			j.File(".npmrc", func(j *J) { j.Content("NEW-NPMRC\n") })
		})
	}); err != nil {
		t.Fatal(err)
	}

	vol := m.Vol()
	for p, want := range map[string]string{
		"/out/p/.env.old":   "OLD-ENV\n",
		"/out/p/.npmrc.old": "OLD-NPMRC\n",
		"/out/p/.env":       "NEW-ENV\n",
		"/out/p/.npmrc":     "NEW-NPMRC\n",
	} {
		if got := string(vol[p]); got != want {
			t.Errorf("%s: got %q want %q", p, got, want)
		}
	}
}

func TestPreserveBackupNamingUnchangedForNormalFiles(t *testing.T) {
	m := NewMemFS()
	_ = m.WriteFile("/out/p/a.txt", []byte("OLD\n"))
	_ = m.WriteFile("/out/p/b.min.js", []byte("OLDJS\n"))

	tr := true
	j := containJ(m, WithExisting(Existing{Txt: ExistingTxt{Preserve: &tr}}))
	if _, err := j.Generate(Options{}, func(j *J) {
		j.Project(ProjectProps{Folder: "p"}, func(j *J) {
			j.File("a.txt", func(j *J) { j.Content("NEW\n") })
			j.File("b.min.js", func(j *J) { j.Content("NEWJS\n") })
		})
	}); err != nil {
		t.Fatal(err)
	}

	vol := m.Vol()
	for p, want := range map[string]string{
		"/out/p/a.old.txt":    "OLD\n",
		"/out/p/b.min.old.js": "OLDJS\n",
	} {
		if got := string(vol[p]); got != want {
			t.Errorf("%s: got %q want %q", p, got, want)
		}
	}
}

func TestAnnotatedPath(t *testing.T) {
	cases := []struct{ in, kind, want string }{
		{"foo/bar.txt", "old", "foo/bar.old.txt"},
		{"foo/bar.txt", "new", "foo/bar.new.txt"},
		{"foo/.env", "old", "foo/.env.old"},
		{"foo/.npmrc", "old", "foo/.npmrc.old"},
		{"foo/noext", "old", "foo/noext.old"},
		{"foo/a.b.txt", "old", "foo/a.b.old.txt"},
		{"bar.txt", "old", "bar.old.txt"},
	}
	for _, c := range cases {
		if got := annotatedPath(c.in, c.kind); got != c.want {
			t.Errorf("annotatedPath(%q,%q) = %q want %q", c.in, c.kind, got, c.want)
		}
	}
}

func TestValidName(t *testing.T) {
	ok := []string{"a.txt", "sub/a.txt", "/code/js", "...", "..foo", "foo..", ""}
	for _, n := range ok {
		if err := validName(n, "File"); err != nil {
			t.Errorf("validName(%q) unexpectedly rejected: %v", n, err)
		}
	}
	bad := []string{"..", "../x", "a/../../x", "..\\x", "sub/.."}
	for _, n := range bad {
		if err := validName(n, "File"); !errors.Is(err, ErrNameTraversal) {
			t.Errorf("validName(%q) should have been rejected, got %v", n, err)
		}
	}
}

File: ts/test/containment.test.ts
// Regression tests for output-path containment and backup-file naming.
//
// These three defects share a theme: a generated path escaping where the
// caller expected it to land, or two files colliding on one backup path.
// All are data-integrity issues, so each keeps an explicit test.

import { test, describe } from 'node:test'
import { expect } from './expect'

import { memfs } from '../dist/util/memfs'

import {
  Jostraca,
  Project,
  Folder,
  File,
  Content,
} from '../'


// 2025-01-01T00:00:00.000Z
const START_TIME = 1735689600000

const outkeys = (vol: any) =>
  Object.keys(vol.toJSON()).filter((k: string) => !k.includes('.jostraca')).sort()


describe('containment', () => {

  // A File with no enclosing Project used to join onto an empty folder
  // path, producing '/<name>' — an absolute path at the filesystem root,
  // ignoring the configured output folder entirely.
  test('file-without-project-stays-in-output-folder', async () => {
    const { fs, vol } = memfs({})
    const jostraca = Jostraca({ now: () => START_TIME, fs: () => fs })

    await jostraca.generate({ folder: '/out' }, () => {
      File({ name: 'x.txt' }, () => Content('hi\n'))
    })

    expect(outkeys(vol)).equal(['/out/x.txt'])
  })


  test('folder-without-project-stays-in-output-folder', async () => {
    const { fs, vol } = memfs({})
    const jostraca = Jostraca({ now: () => START_TIME, fs: () => fs })

    await jostraca.generate({ folder: '/out' }, () => {
      Folder({ name: 'sub' }, () => {
        File({ name: 'y.txt' }, () => Content('hi\n'))
      })
    })

    expect(outkeys(vol)).equal(['/out/sub/y.txt'])
  })


  // Names compose straight into output paths and models are routinely
  // third-party data, so a `..` segment is an arbitrary-file-write.
  test('file-name-rejects-traversal', async () => {
    const { fs } = memfs({})
    const jostraca = Jostraca({ now: () => START_TIME, fs: () => fs })

    await expect(async () =>
      await jostraca.generate({ folder: '/out' }, () =>
        Project({ folder: 'p' }, () =>
          File({ name: '../../../../etc/pwned.txt' }, () => Content('OWNED\n')))))
      .rejects(/must not contain a "\.\." path segment/)
  })


  test('folder-name-rejects-traversal', async () => {
    const { fs } = memfs({})
    const jostraca = Jostraca({ now: () => START_TIME, fs: () => fs })

    await expect(async () =>
      await jostraca.generate({ folder: '/out' }, () =>
        Project({ folder: 'p' }, () =>
          Folder({ name: '../..' }, () =>
            File({ name: 'e.txt' }, () => Content('X\n'))))))
      .rejects(/must not contain a "\.\." path segment/)
  })


  // A leading `/` in a Folder name is a supported feature (it composes
  // with the Project folder) and must keep working — see the
  // `absolute_paths` parity scenario.
  test('absolute-folder-name-still-composes', async () => {
    const { fs, vol } = memfs({})
    const jostraca = Jostraca({ now: () => START_TIME, fs: () => fs })

    await jostraca.generate({ folder: '/top' }, () => {
      Project({ folder: '/top/sdk' }, () => {
        Folder({ name: '/code/js' }, () => {
          File({ name: 'foo.js' }, () => Content('// foo\n'))
        })
      })
    })

    expect(outkeys(vol)).equal(['/top/sdk/code/js/foo.js'])
  })


  // Node's Path.extname('.env') is '', so the old regex-strip removed the
  // whole name: every dotfile in a folder collapsed onto the same `.old`
  // backup and the second one destroyed the first one's copy.
  test('preserve-backs-up-dotfiles-distinctly', async () => {
    const { fs, vol } = memfs({})
    fs.mkdirSync('/out/p', { recursive: true })
    fs.writeFileSync('/out/p/.env', 'OLD-ENV\n')
    fs.writeFileSync('/out/p/.npmrc', 'OLD-NPMRC\n')

    const jostraca = Jostraca({ now: () => START_TIME, fs: () => fs })

    await jostraca.generate(
      { folder: '/out', existing: { txt: { preserve: true } } },
      () => Project({ folder: 'p' }, () => {
        File({ name: '.env' }, () => Content('NEW-ENV\n'))
        File({ name: '.npmrc' }, () => Content('NEW-NPMRC\n'))
      }))

    const json: any = vol.toJSON()
    expect(json['/out/p/.env.old']).equal('OLD-ENV\n')
    expect(json['/out/p/.npmrc.old']).equal('OLD-NPMRC\n')
    expect(json['/out/p/.env']).equal('NEW-ENV\n')
    expect(json['/out/p/.npmrc']).equal('NEW-NPMRC\n')
  })


  // Backup naming for ordinary files must be unchanged by the dotfile fix.
  test('preserve-backup-naming-unchanged-for-normal-files', async () => {
    const { fs, vol } = memfs({})
    fs.mkdirSync('/out/p', { recursive: true })
    fs.writeFileSync('/out/p/a.txt', 'OLD\n')
    fs.writeFileSync('/out/p/b.min.js', 'OLDJS\n')

    const jostraca = Jostraca({ now: () => START_TIME, fs: () => fs })

    await jostraca.generate(
      { folder: '/out', existing: { txt: { preserve: true } } },
      () => Project({ folder: 'p' }, () => {
        File({ name: 'a.txt' }, () => Content('NEW\n'))
        File({ name: 'b.min.js' }, () => Content('NEWJS\n'))
      }))

    const json: any = vol.toJSON()
    expect(json['/out/p/a.old.txt']).equal('OLD\n')
    expect(json['/out/p/b.min.old.js']).equal('OLDJS\n')
  })

})
